src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -45,30 +75,49 @@ public override string ToString()
                 CharacterIndex);
         }
 
+        /// <inheritdoc />
         public override bool Equals(object obj)
         {
             return (obj is SourceLocation) && Equals((SourceLocation)obj);
         }
 
+        /// <inheritdoc />
         public override int GetHashCode()
         {
             // LineIndex and CharacterIndex can be calculated from AbsoluteIndex and the document content.
-            return AbsoluteIndex;
+            return HashCodeCombiner.Start().Add(FilePath).Add(AbsoluteIndex).CombinedHash;
         }
 
+        /// <inheritdoc />
         public bool Equals(SourceLocation other)
         {
-            return 
+            return string.Equals(FilePath, other.FilePath, StringComparison.OrdinalIgnoreCase) &&
```

Comment:
also this `Equals()` doesn't match `GetHashCode()`: source locations for file `Hello.world` may equal those for `HELLO.world` but their hash codes will be different.  must either change `GetHashCode()` to use `Add(filePath, StringComparison.OrdinalIgnoreCase)` or use `StringComparison.Ordinal` here.  not sure which is the better choice given oddities in lower layers w.r.t. case sensitivity.